### PR TITLE
Implement CopyObjectPart API

### LINKED ISSUE #3662: Latest docker registry fails to work with Minio
```
time="2017-01-30T19:53:59Z" level=info msg="response completed" go.version=go1.7.3 http.request.host="localhost:9001" http.request.id=400d202c-abae-45a1-8734-abde39ec7633 http.request.method=PATCH http.request.remoteaddr="172.17.0.1:33798" http.request.uri="/v2/myfirstimage/blobs/uploads/ac626b47-11a3-4cff-9cb8-0ae766500b51?_state=oyAvPtiJxO1k4o7d7xToF-xRpD6mLOTgvHC6DM3VhKR7Ik5hbWUiOiJteWZpcnN0aW1hZ2UiLCJVVUlEIjoiYWM2MjZiNDctMTFhMy00Y2ZmLTljYjgtMGFlNzY2NTAwYjUxIiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE3LTAxLTMwVDE5OjUzOjM2LjcxMzM0MzUxNloifQ%3D%3D" http.request.useragent="docker/1.13.0 go/go1.7.3 git-commit/49bf474 kernel/4.4.0-59-generic os/linux arch/amd64 UpstreamClient(Docker-Client/1.13.0 \\(linux\\))" http.response.duration=22.154395686s http.response.status=202 http.response.written=0 instance.id=f1cf2142-9797-4e3a-b5d9-08e1321f51b8 version=v2.6.0 
172.17.0.1 - - [30/Jan/2017:19:53:37 +0000] "PATCH /v2/myfirstimage/blobs/uploads/ac626b47-11a3-4cff-9cb8-0ae766500b51?_state=oyAvPtiJxO1k4o7d7xToF-xRpD6mLOTgvHC6DM3VhKR7Ik5hbWUiOiJteWZpcnN0aW1hZ2UiLCJVVUlEIjoiYWM2MjZiNDctMTFhMy00Y2ZmLTljYjgtMGFlNzY2NTAwYjUxIiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE3LTAxLTMwVDE5OjUzOjM2LjcxMzM0MzUxNloifQ%3D%3D HTTP/1.1" 202 0 "" "docker/1.13.0 go/go1.7.3 git-commit/49bf474 kernel/4.4.0-59-generic os/linux arch/amd64 UpstreamClient(Docker-Client/1.13.0 \\(linux\\))"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4a8122]

goroutine 303 [running]:
panic(0xb96fe0, 0xc420012050)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).copy.func1(0xc4203aa420, 0x0, 0xc420315c00, 0x1066a40, 0xc4202c7d00, 0xc420222b40, 0x60, 0xc4203c4ee0, 0x69, 0xc4203a21e0, ...)
	/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:742 +0x592
created by github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).copy
	/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:748 +0x6aa
```

## Expected Behavior
Docker registry should work fine with Minio server.

## Current Behavior
Latest docker registry release does not work one has to use 2.4 or 2.5

## Steps to Reproduce (for bugs)

```
$ cat Dockerfile 
FROM registry:2.5
COPY config.yml /etc/docker/registry/config.yml
$ cat config.yml 
version: 0.1
log:
    fields:
        service: registry
http:
    addr: :5000
storage:
    cache:
        layerinfo: inmemory
    s3:
        accesskey: Q3AM3UQ867SPQQA43P2F
        secretkey: zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG
        region: us-east-1
        regionendpoint: https://play.minio.io:9000/
        bucket: docker-registry
        encrypt: false
        secure: true
        v4auth: true
        chunksize: 5242880
        rootdirectory: /

$ docker build . --tag=my-registry
$ docker run -p 9001:5000 my-registry
```

On another terminal 

```
$ docker tag ubuntu localhost:9001/myfirstimage
$ docker push localhost:9001/myfirstimage
```

## Context
To use docker registry on Minio.

## Your Environment
docker 1.13
minio master and latest release.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
This API is implemented to allow copying data from an
existing source object to an ongoing multipart operation

http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html

## Motivation and Context
Fixes #3662

## How Has This Been Tested?
Manually using docker registry.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/fs-v1-multipart.go [hunk 1]
```diff
@@ -451,6 +451,42 @@ func partToAppend(fsMeta fsMetaV1, fsAppendMeta fsMetaV1) (part objectPartInfo,
...
+		pipeWriter.Close() // Close writer explicitly signalling we wrote all data.
+	}()
+
+	md5Sum, err := fs.PutObjectPart(dstBucket, dstObject, uploadID, partID, length, pipeReader, "", "")
+	if err != nil {
+		return "", time.Time{}, toObjectErr(err, dstBucket, dstObject)
+	}
+
+	// Explicitly close the reader.
+	pipeReader.Close()
```
Comment (krisis) [inline cmd/fs-v1-multipart.go:481]: Shouldn't we call ``pipeReader.Close()`` even when ``fs.PutObjectPart`` fails?

File: cmd/fs-v1-multipart.go [hunk 2]
```diff
@@ -451,6 +451,42 @@ func partToAppend(fsMeta fsMetaV1, fsAppendMeta fsMetaV1) (part objectPartInfo,
...
+
+	return md5Sum, time.Now().UTC(), nil
```
Comment (krisis) [inline cmd/fs-v1-multipart.go]: Shouldn't the time returned be identical to the modTime set in destination object's xl.json?

File: cmd/xl-v1-multipart.go [hunk 3]
```diff
@@ -538,6 +539,45 @@ func (xl xlObjects) NewMultipartUpload(bucket, object string, meta map[string]st
...
+		pipeWriter.Close() // Close writer explicitly signalling we wrote all data.
+	}()
+
+	md5Sum, err := xl.PutObjectPart(dstBucket, dstObject, uploadID, partID, length, pipeReader, "", "")
+	if err != nil {
+		return "", time.Time{}, toObjectErr(err, dstBucket, dstObject)
+	}
+
+	// Explicitly close the reader.
+	pipeReader.Close()
```
Comment (krisis) [inline cmd/xl-v1-multipart.go:571]: Shouldn't we call ``pipeReader.Close()`` even when ``xl.PutObjectPart`` fails?

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/fs-v1-multipart.go:481]: there's no downside in not closing it when putobjectpart fails, i think.

**Review by krishnasrinivas — APPROVED** [review]
Looks good, but KP's comment about returning xl.json modTime needs to be addressed though.

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (krisis) [reply cmd/fs-v1-multipart.go:481]: @krishnasrinivas could you explain what happens if we don't? How and when is the pipe (fd) closed?

Comment (harshavardhana) [reply cmd/fs-v1-multipart.go:481]: Writer close is sufficient reader close is just for brevity it is not really needed to be closed..

File: cmd/fs-v1-multipart.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/fs-v1-multipart.go]: This is correct.

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (krisis) [reply cmd/fs-v1-multipart.go:481]: OK.

**Review by krisis — APPROVED** [review]

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/fs-v1-multipart.go:481]: @krisis pipes will be like channles, i.e it gets garbage collected when there are no refs. And if there are no pipe readers then doing pipeWriter.Close() will not have any advantage.

**Pull request merged**